File: src/game/systems/death.lua
local Aspect = require 'ecs.aspect'
local System = require 'ecs.system'
local Fixture = require 'game.components.fixture'
local Health = require 'game.components.health'
local Player = require 'game.components.player'
local Respawn = require 'game.components.respawn'

local aspect = Aspect.new({Health})
local Death = System:new('death', aspect)

local function destroyJoints(entity)
  local fixture = entity:as(Fixture)
  local joints = fixture.fixture:getBody():getJoints()
  for _, joint in pairs(joints) do
    local a, b = joint:getBodies()
    a:getUserData().entity:destroy()
    b:getUserData().entity:destroy()
  end
end

function Death:container(dt, entity)
  entity:destroy()
end

function Death:mob(dt, entity)
  destroyJoints(entity)
  entity:destroy()
end

function Death:player(dt, entity)
  local health = entity:as(Health)
  local player = entity:as(Player)

  entity:add(Respawn.new(1))
  health.hitpoints = 1
end

function Death:update(dt)
  for _, entity in pairs(self.entities) do
    local health = entity:as(Health)

    if (health.hitpoints <= 0) then
      local fn = self[entity.meta.type]

      if (fn) then
        fn(self, dt, entity)
      end
    end
  end
end

return Death

File: packages/client/src/game/systems/Death.ts
import { System } from '../../ecs/System'
import { SystemFlag } from '../flags'
import { Aspect } from '../../ecs/Aspect'
import { Health } from '../components/Health'
import { Respawn } from '../components/Respawn'
import { Player } from '../components/Player'
import { Blueprint } from '../utils/factory'
import { Entity } from '../../ecs/Entity'

const deathFunctions: { [key: string]: (entity: Entity) => void } = {
  [Blueprint.Player]: (entity: Entity) => {
    const health = entity.as(Health)
    const player = entity.as(Player)

    if (!player || !health) {
      return
    }

    player.lives--

    entity.add(new Respawn(3))
  },
}

export class DeathSystem extends System {
  static _id = 'DeathSystem'
  _id = DeathSystem._id

  static _flag = SystemFlag.Death
  _flag = DeathSystem._flag

  static _aspect = new Aspect([Health])
  _aspect = DeathSystem._aspect

  update = (dt: number) => {
    this.entities.forEach(entity => {
      const health = entity.as(Health)

      if (!health) {
        return
      }

      if (health.hitpoints <= 0 && entity.userData.blueprint) {
        const fn = deathFunctions[entity.userData.blueprint]

        fn && fn(entity)
      }
    })
  }
}
